src/error/go.rs

```diff
@@ -86,7 +95,11 @@ impl GoResult {
                 // that means it was initialized by the go code, with values generated by `memory::allocate_rust`
                 String::from_utf8_lossy(&output.consume()).into()
             };
-            Err(FfiError::unknown(msg))
+            if is_user_error {
+                Err(FfiError::user_err(msg))
+            } else {
+                Err(FfiError::unknown(msg))
+            }
```

Comment:
Looks correct, but this is getting complicated because now this implementation needs to know for which cases `self.try_into()` fails. I'd avoid relying on `self.try_into()` and instead copy the match statement from `fn try_from(other: GoResult) -> Result<Self, Self::Error> {` in here:

```rust
    pub unsafe fn into_ffi_result<F>(self, output: Buffer, default: F) -> Result<(), FfiError>
    where
        F: Fn() -> String,
    {
        let msg = if output.ptr.is_null() {
            default()
        } else {
            // We initialize `output` with a null pointer. if it is not null,
            // that means it was initialized by the go code, with values generated by `memory::allocate_rust`
            String::from_utf8_lossy(&output.consume()).into()
        };

        match self {
            GoResult::Ok => Ok(()),
            GoResult::Panic => Err(FfiError::foreign_panic()),
            GoResult::BadArgument => Err(FfiError::bad_argument()),
            GoResult::OutOfGas => Err(FfiError::out_of_gas()),
            GoResult::Other => Err(FfiError::unknown(msg)),
            GoResult::User => Err(FfiError::user_err(msg)),
        }
    }
```

Reply:
Will do